### PR TITLE
[WIP] Take customer group from logged in customer instead of from quote, qu…

### PR DESCRIPTION
**Pull request by hostep**
…ote sometimes seems to have the wrong customer group id if switching billing addresses a few times in the payment step of the checkout.

### Description (*)
Sorry, no description yet, just a PR to run all the tests to see what the outcome will be.
I have the feeling after some testing that this fixes an edge case bug, but might be wrong ...

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hostep** [comment]
@magento run all tests

**Comment by hostep** [comment]
Closing because this is not the correct solution. Guest users have the same problem.
Totals on the quote are calculated based on the custom group id from the quote data in the database which might be outdated, because the customer group id that might have changed by picking a different address in the frontend.

A more correct solution would be that the custom group id is set to the quote **before** totals/tax calculation starts happening. We have the impression this is set later.

A colleague of mine is investigating to see if there is a better solution.

**Comment by hostep** [comment]
Issue is related to sort order in which addresses are processed in php to calculate totals.

For tax calculations in EU countries, the billing address should get processed first, so the correct customer group is calculated, to see if this is a B2C or B2B sale, before the shipping address should get processed to calculate how much tax should be applied.

We have an experimental patch like this that seems to fix our issue (applied to quote module):
```diff
--- a/Model/Quote/TotalsCollector.php
+++ b/Model/Quote/TotalsCollector.php
@@ -143,8 +143,14 @@ class TotalsCollector
         $total->setGrandTotal(0);
         $total->setBaseGrandTotal(0);

+        $allQuoteAddresses = $quote->getAllAddresses();
+        // Sort addresses to make sure that billing address is first
+        usort($allQuoteAddresses, function ($a, $b) {
+            return $a->getAddressType() == 'billing' ? -1 : 1;
+        });
+
         /** @var \Magento\Quote\Model\Quote\Address $address */
-        foreach ($quote->getAllAddresses() as $address) {
+        foreach ($allQuoteAddresses as $address) {
             $addressTotal = $this->collectAddressTotals($quote, $address);

             $total->setShippingAmount($addressTotal->getShippingAmount());
```

A better fix for Magento core, is probably to look at the configured type of address in 'Customers > Customer Configuraton > Create New Account Options > Tax Calculation Based On', and always sort that one first to be processed
(mind that this config only shows up when 'Enable Automatic Assignment to Customer Group' is enabled)
